skbio/metadata/_interval.py

```diff
@@ -0,0 +1,756 @@
+                raise TypeError(
+                    'Cannot give a non-iterable (%r) '
+                    'to `boundaries`.' % boundaries)
+
+            if len(boundaries) != spans:
+                raise ValueError(
+                    'The length of boundaries must '
+                    'be equal to the length of locations {!r}.')
+
+            for boundary in boundaries:
+                _assert_valid_boundary(boundary)
+
+        if locations is None:
+            # `locations` and `boundaries` cannot both be omitted.
+            if boundaries is None:
+                raise ValueError('You must give `None` to both `locations` '
+                                 'and `boundaries`.')
+            # If only `boundaries` is provided, set `self.boundaries` and don't
+            # change `self.locations`.
+            else:
+                self._boundaries = boundaries
+        else:
+            # If only `locations` is provided, reset `self.boundaries` to
+            # all `True`.
+            if boundaries is None:
+                locations.sort()
+                self._locations = locations
+                # reset all the boundaries to True!!
+                del self.boundaries
+
+            # If both `locations` and `boundaries` are provided, set
+            # `self.locations` and `self.boundaries`.
+            else:
+                combo = [i for i in sorted(zip(locations, boundaries))]
+                # use the boundaries setter to validate the value
+                self._boundaries = [i[1] for i in combo]
+                self._locations = [i[0] for i in combo]
+
+            self._interval_metadata._is_stale_tree = True
+
+    @property
+    @experimental(as_of='0.5.0-dev')
+    def boundaries(self):
+        '''The openness of each coordinate.
+
+        This indicates that the exact boundary point of a interval feature
+        is unknown. The location may begin or end at some points outside
+        the specified coordinates. This accommodates the location format [1]_
+        of INSDC.
+
+        References
+        ----------
+        .. [1] ftp://ftp.ebi.ac.uk/pub/databases/embl/doc/FT_current.html#3.4.3
+        '''
+        return self._boundaries
+
+    @boundaries.setter
+    @experimental(as_of='0.5.0-dev')
+    def boundaries(self, value):
+        '''Set ``boundaries``.
+
+        The ``value`` should an iterable matching ``self.locations``.
+        '''
+        self._locations_boundaries_setter(boundaries=value)
+
+    @boundaries.deleter
+    @experimental(as_of='0.5.0-dev')
+    def boundaries(self):
+        '''Delete ``boundaries``.
+
+        This set all boundaries to be True.
+        '''
+        if self.dropped:
+            raise RuntimeError('Cannot change boundaries dropped '
+                               'Interval object.')
+        self._boundaries = [(True, True)] * len(self.locations)
+
+    @property
+    @experimental(as_of='0.5.0-dev')
+    def locations(self):
```

Comment:
Haha I really need this 3 day weekend :P

On Friday, September 2, 2016, Zech Xu <notifications@github.com> wrote:

> In skbio/metadata/_interval.py
> <https://github.com/biocore/scikit-bio/pull/1414#discussion_r77428986>:
>
> > +
> > +    @boundaries.deleter
> > +    @experimental(as_of='0.5.0-dev')
> > +    def boundaries(self):
> > +        '''Delete ``boundaries``.
> > +
> > +        This set all boundaries to be True.
> > +        '''
> > +        if self.dropped:
> > +            raise RuntimeError('Cannot change boundaries dropped '
> > +                               'Interval object.')
> > +        self._boundaries = [(True, True)] * len(self.locations)
> > +
> > +    @property
> > +    @experimental(as_of='0.5.0-dev')
> > +    def locations(self):
>
> No worries. I know it is Friday afternoon :D On Fri, Sep 2, 2016 at 6:28
> PM, Jai Ram Rideout <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
> … <#m_-583954836256589134_>
> In skbio/metadata/_interval.py <#1414 (comment)
> <https://github.com/biocore/scikit-bio/pull/1414#discussion_r77428767>>:
> > + > + @boundaries.deleter > + @experimental
> <https://github.com/experimental>(as_of='0.5.0-dev') > + def
> boundaries(self): > + '''Delete ``boundaries``. > + > + This set all
> boundaries to be True. > + ''' > + if self.dropped: > + raise
> RuntimeError('Cannot change boundaries dropped ' > + 'Interval object.') >
> + self._boundaries = [(True, True)] * len(self.locations) > + > +
> @property <https://github.com/property> > + @experimental
> <https://github.com/experimental>(as_of='0.5.0-dev') > + def
> locations(self): Yeah the current code is correct for these attributes.
> Sorry I didn't confirm that. I really confused myself with the behavior,
> enough that I didn't even notice it was correct! Thanks for your patience
> on this one :) … <#m_2071656106224454884_> On Friday, September 2, 2016,
> Zech Xu ***@***.***> wrote: In skbio/metadata/_interval.py <#1414
> <https://github.com/biocore/scikit-bio/pull/1414> (comment) <#1414
> (comment)
> <https://github.com/biocore/scikit-bio/pull/1414#discussion_r77427388>>>:
> > + > + @boundaries.deleter > + @experimental
> <https://github.com/experimental> <https://github.com/
> experimental>(as_of='0.5.0-dev') > + def boundaries(self): > + '''Delete
> ``boundaries``. > + > + This set all boundaries to be True. > + ''' > + if
> self.dropped: > + raise RuntimeError('Cannot change boundaries dropped ' >
> + 'Interval object.') > + self._boundaries = [(True, True)] *
> len(self.locations) > + > + @property <https://github.com/property> <
> https://github.com/property> > + @experimental
> <https://github.com/experimental> <https://github.com/
> experimental>(as_of='0.5.0-dev') > + def locations(self): agree on your
> comments and it is implemented as you said here, right? except for missing
> the deleter for locations. locations should always have a meaningful value.
> so no deleter for it. — You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <
> https://github.com/biocore/scikit-bio/pull/1414/files/ 1837b1b
> <https://github.com/biocore/scikit-bio/commit/1837b1b85f84a93c77811f3c1a425f7da63c4a69>#r77427388>,
> or mute the thread < https://github.com/notifications/unsubscribe-
> auth/ABwvwJjQU63-Nh0DgdK- 1v3sz9UAPzrOks5qmL6TgaJpZM4JoIbD> . -- Sent
> from my phone. — You are receiving this because you were mentioned. Reply
> to this email directly, view it on GitHub <https://github.com/biocore/
> scikit-bio/pull/1414/files/1837b1b85f84a93c77811f3c1a425f
> 7da63c4a69#r77428767>, or mute the thread <https://github.com/
> notifications/unsubscribe-auth/AA9CeHnfj6FivytlOoLa2LIrWNleys
> Ddks5qmM1agaJpZM4JoIbD> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1414/files/1837b1b85f84a93c77811f3c1a425f7da63c4a69#r77428986>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABwvwIh4GYGi5sL28qecY4W4epsDziyWks5qmM_GgaJpZM4JoIbD>
> .
>


-- 
Sent from my phone.